### PR TITLE
Delete action

### LINKED ISSUE #443: #20 We should be able to delete an action
Q: Which users/roles should be able to delete

### PR DESCRIPTION
**Pull request by achaikou**
closes #443 

It still fails randomly asserting that deleted action is found in DOM, even though it is clearly deleted. Cypress waits for state to change, so reason for the failure is unclear.
Might be that it will behave more stable on CI, then locally, but shall see.

Upd: regarding Codacy.
Looks like .eslintrc.json is not picked up by codacy at all while it works locally with eslint.

It might be that Codacy just doesn't pick the files up. We might have to explicitly define instead all the rules we need (and Codacy lacks export rules feature, so it might be painful as we don't exactly know what settings were used), put them into eslint file in root directory, merge, and then request Codacy to take rules from config file itself. In that case they will have to be common for all the directories.
If that's so, Codacy settings will be taken as a separate PR.

Our other options are:
- explicit comments (it works, by would be tiring) 
- disable the no-undef check completely in UI (which should be fast and easy). After all, we do typescript which should pick up undeclared variables on its own

### CODE REVIEW

File: backend/api/GQL/Mutation.cs [hunk 1]
```diff
@@ -185,6 +185,16 @@ public Action EditAction(string actionId, string assignedToId, string descriptio
             return _actionService.EditAction(action, assignedTo, description, dueDate, title, onHold, completed, priority);
         }
 
+        public Action DeleteAction(string actionId)
+        {
+            /* Note that no related fields are loaded */
+            IQueryable<Action> queryableAction = _actionService.GetAction(actionId);
+            Action action = queryableAction.First();
```
Comment (ErlendHaa) [inline backend/api/GQL/Mutation.cs:192]: IQueriable.First throws InvalidOperationException if sequence is empty [1]

[1] https://docs.microsoft.com/en-us/dotnet/api/system.linq.queryable.first?view=net-5.0

File: frontend/src/api/mutations.ts [hunk 2]
```diff
@@ -150,6 +150,59 @@ export const useEditActionMutation = (): EditActionMutationProps => {
...
+        update(cache, mutationResult) {
+            const deleteAction = mutationResult.data.deleteAction
+            cache.modify({
+                id: cache.identify({
+                    __typename: 'Question',
+                    id: questionId,
+                }),
+                fields: {
+                    actions(existingActionRefs, { readField }) {
+                        return existingActionRefs.filter((actionRef: any) => {
```
Comment (ErlendHaa) [inline frontend/src/api/mutations.ts:185]: I guess the apollo cache does not cascade deletions? We probably have dangling notes in the cache. Probably not worth fixing?

File: frontend/src/components/Action/QuestionActionsList.tsx [hunk 3]
```diff
@@ -98,6 +119,19 @@ const QuestionActionsList = ({ question, participants }: Props) => {
                 connectedQuestion={question}
                 possibleAssignees={participants}
             />
+            <ConfirmationDialog
+                isOpen={isConfirmDeleteDialogOpen}
+                title="Delete Action?"
+                description="Deleting an action will remove it and its notes from Workshop and Follow-up tabs."
```
Comment (ErlendHaa) [inline frontend/src/components/Action/QuestionActionsList.tsx:126]: `... will permanently delete it ... from the evaluation.` ?

File: frontend/cypress.json [hunk 4]
```diff
@@ -1 +1,3 @@
-{}
+{
+    "chromeWebSecurity": false
```
Comment (ErlendHaa) [inline frontend/cypress.json:2]: Why?

File: frontend/cypress/integration/actions_spec.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+            seed.addAction(actionToDelete).addAction(actionToStay)
+            return { seed, actionToDelete, actionToStay }
+        }
+
+        let seed: EvaluationSeed
+        let actionToDelete: Action
+        let actionToStay: Action
+
+        beforeEach(() => {
+            ;({ seed, actionToDelete, actionToStay } = createSeed())
```
Comment (ErlendHaa) [inline frontend/cypress/integration/actions_spec.ts:37]: `const { seed, actionToDelete, actionToStay } = createSeed()` ?

File: frontend/cypress/integration/actions_spec.ts [hunk 6]
```diff
@@ -0,0 +1,97 @@
+import { EvaluationSeed } from '../support/evaluation_seed'
+import { Priority, Progression } from '../../src/api/models'
+import { Action } from '../support/mocks'
+import ActionsGrid from '../support/action_grid'
+import { ConfirmationDialog } from '../support/common'
+import { DELETE_ACTION } from '../support/gql'
+
+describe('Actions', () => {
+    context('Delete', () => {
+        const createSeed = () => {
```
Comment (ErlendHaa) [inline frontend/cypress/integration/actions_spec.ts:11]: Should we decide on where to define such helpers? I would opt for putting everything that doesn't _need_ to be inside a describe, outside of it. Then the actual test cases pop a bit more

File: backend/api/GQL/Mutation.cs (hunk 1 above)
Comment (achaikou) [reply backend/api/GQL/Mutation.cs:192]: Yep, but code around is also using First, not FirstOrDefault?

Do you have a better suggestion what to use here? If element is not available we want to fail? So why not to use the default error?

File: frontend/cypress.json (hunk 4 above)
Comment (achaikou) [reply frontend/cypress.json:2]: We get an "Cypress detected a cross origin error happened" error when reloading the page. I assume fusion tries to do something under the hood
https://docs.cypress.io/guides/guides/web-security#Limitations

File: frontend/cypress/integration/actions_spec.ts (hunk 6 above)
Comment (achaikou) [reply frontend/cypress/integration/actions_spec.ts]: Agreed to wait with precise agreement, but will move createSeed below all the tests, as contents of databse bar returned values are not important

File: frontend/src/api/mutations.ts (hunk 2 above)
Comment (achaikou) [reply frontend/src/api/mutations.ts:185]: We have Notes and deleted Action still in cache, and they are not removed by cache.gc().
No idea why, but at least both objects are not visible. Maybe I misunderstand, and Action and Note are actually also considered root objects, and must be removed with evict?
No idea how they ended up there as well.

Maybe that's something we'll be able to solve with more caching experience :woman_shrugging:

**Comment by ReedOnly** [comment]
Regarding .eslintrc.json, we have to manually enable it in Codacy settings to read that file, but then it will _only_ follow the rules from the config file, and it looks like the file need to be in repository root as well. I suggest that we disable no undef check for frontend, because like you said Typescript will inform about that anyways.

File: frontend/cypress/support/mocks.ts [hunk 7]
```diff
@@ -61,47 +51,64 @@ export class Answer {
...
+        createdBy = faker.random.arrayElement(seed!.participants),
+        // no access to questions as creation is run before plant. Reconsider?
+        questionOrder = faker.datatype.number({ min: 1, max: 3 }),
+        createDate = faker.date.past(),
+        dueDate = faker.date.future(),
+        title = faker.lorem.sentence(),
+        priority = faker.random.arrayElement(Object.values(Priority)),
+        description = faker.lorem.words(),
+        completed = faker.datatype.boolean(),
+        onHold = faker.datatype.boolean(),
```
Comment (ErlendHaa) [inline frontend/cypress/support/mocks.ts:97]: I would argue it's a bit weird to pass in seed here with the sole purpose of setting default values. Would suggest giving this responsibility to the createAction function, and leave the constructor as it was

**Review by ErlendHaa — APPROVED** [review]

**Pull request merged**